api/pkg/artifact/artifact.go

```diff
@@ -133,24 +176,135 @@ func (gac *GcsArtifactClient) DeleteArtifact(ctx context.Context, url string) er
 	return nil
 }
 
+type S3ArtifactClient struct {
+	URLScheme
+	client *s3.Client
+}
+
+func NewS3ArtifactClient() (Service, error) {
+	cfg, err := config.LoadDefaultConfig(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("%s,failed loading s3 config for the artifact client", err.Error())
+	}
+	client := s3.NewFromConfig(cfg, func(o *s3.Options) {
+		o.BaseEndpoint = aws.String(os.Getenv("AWS_ENDPOINT_URL"))
+	})
+	return &S3ArtifactClient{
+		URLScheme: s3URLScheme,
+		client:    client,
+	}, nil
+}
+
+func (s3c *S3ArtifactClient) GetType() string {
+	return s3ArtifactClientType
+}
+
+func (s3c *S3ArtifactClient) ReadArtifact(ctx context.Context, url string) ([]byte, error) {
+	u, err := s3c.ParseURL(url)
+	if err != nil {
+		return nil, err
+	}
+
+	reader, err := s3c.client.GetObject(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(u.Bucket),
+		Key:    aws.String(u.Object),
+	})
+	if err != nil {
+		var nsk *types.NoSuchKey
+		if errors.As(err, &nsk) {
+			return nil, ErrObjectNotExist
+		}
+		return nil, err
+	}
+	defer reader.Body.Close() //nolint:errcheck
+
+	bytes, err := io.ReadAll(reader.Body)
+	if err != nil {
+		return nil, err
+	}
+	return bytes, nil
+}
+
+func (s3c *S3ArtifactClient) WriteArtifact(ctx context.Context, url string, content []byte) error {
+	u, err := s3c.ParseURL(url)
+	if err != nil {
+		return err
+	}
+
+	_, err = s3c.client.PutObject(ctx, &s3.PutObjectInput{
+		Bucket: aws.String(u.Bucket),
+		Key:    aws.String(u.Object),
+		Body:   bytes.NewReader(content),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s3c *S3ArtifactClient) DeleteArtifact(ctx context.Context, url string) error {
+	u, err := s3c.ParseURL(url)
+	if err != nil {
+		return err
+	}
+
+	out, err := s3c.client.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
+		Bucket: aws.String(u.Bucket),
+		Prefix: aws.String(u.Object),
+	})
+	if err != nil {
+		return err
+	}
+
+	// TODO: To confirm and refactor if versioning is enabled on S3 and to specify the versionId to be deleted
+	var objects []types.ObjectIdentifier
+	for _, item := range out.Contents {
+		objects = append(objects, types.ObjectIdentifier{
+			Key: item.Key,
+		})
+	}
+	if objects != nil {
+		_, err = s3c.client.DeleteObjects(ctx, &s3.DeleteObjectsInput{
+			Bucket: aws.String(u.Bucket),
+			Delete: &types.Delete{
+				Objects: objects,
+				Quiet:   aws.Bool(true),
+			},
+		})
+		if err != nil {
+			return err
+		}
```

Comment:
This function doesn't return an explicit error if the given objects do not exist. It does however, return an explicit error if the bucket specified cannot be found, but this would be unexpected because an incorrect bucket would've been flagged out by the `ListObjectsV2` call above.